### DISCUSSION TITLE
How much can Mojo be a superset of Python?

### DISCUSSION THREAD

**Discussion by ravinsharma7** [modular/modular#124] [General]
(Food for thought)
This concept of superset languages often leads to difficulties in balancing the introduction of new features in the sub-language with those in the superset language.

This reminds me of multiple open-source stories where feature parity eventually failed to keep up, or the superset language decided to stop maintaining feature parity. 

There are many stories like these in the programming language world, where maintaining feature parity is halted due to difficulties in balancing the performance and criteria of the superset with the sub-language, and the timely update of the superset language to support the sub-language.

I'm curious about Mojo and Modular's philosophy and approach on this.

**Comment by ego** [comment]
> (Food for thought) This concept of superset languages often leads to difficulties in balancing the introduction of new features in the sub-language with those in the superset language.
> 
> This reminds me of multiple open-source stories where feature parity eventually failed to keep up, or the superset language decided to stop maintaining feature parity.
> 
> There are many stories like these in the programming language world, where maintaining feature parity is halted due to difficulties in balancing the performance and criteria of the superset with the sub-language, and the timely update of the superset language to support the sub-language.
> 
> I'm curious about Mojo and Modular's philosophy and approach on this.

I think the main idea is to cover/embrace the Python development space and reuse the huge amount of existing libraries, especially for AI/ML. Without it, it will be difficult to give something new that is useful to the community and industry.
Python has a strong grip and influence in the scientific and engineering space.

**Comment by lattner** [comment]
I have a lot of experience with this, as does the industry, e.g. C++ is a "superset" of C, even though there are tiny differences at the cracks.  We'll build this out over time.  We have FAR more implementation options available to us than any previous python-family project.  We can even provide full compatibility with the existing C APIs if we choose to.  It is all about how we prioritize things over time.
